### PR TITLE
prov/efa: Emulated Atomic Support for Cuda Devices

### PR DESCRIPTION
**Pull request by a-szegel**
Changes to prov/efa to add emulated atomic support for CUDA devices:

1. Added result + fetch memory descriptors to rxr_atomic_ex
2. Switched every memcpy to be ofi_hmem memcpy
3. On the target side, added ofi_hmem memcpy's around emulated atomic impl functions, and performed them on host memory.

Testing:
```
fi_rdm_atomic -D cuda -p efa -E  # passes, and exits cleanly
fi_rdm_atomic -p efa -E                # passes, and exits cleanly
```

Tested both sides no cuda, both sides cuda, and one side cuda, one side host.  There is a big slow down when using atomics with CUDA devices (which is expected).

### CODE REVIEW

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 1]
```diff
@@ -2280,9 +2292,16 @@ ssize_t rxr_pkt_init_compare_rta(struct rxr_ep *ep, struct rxr_op_entry *tx_entr
...
 	 */
 	data = (char *)pkt_entry->pkt + pkt_entry->pkt_size;
-	data_size = ofi_copy_from_iov(data, ep->mtu_size - pkt_entry->pkt_size,
-				      tx_entry->atomic_ex.comp_iov,
-				      tx_entry->atomic_ex.comp_iov_count, 0);
+
+    hmem_iface = ((struct efa_mr*) tx_entry->atomic_ex.compare_desc)->peer.iface;
+    device = ((struct efa_mr*) tx_entry->atomic_ex.compare_desc)->peer.device.reserved;
+
+	data_size = ofi_copy_from_hmem_iov(data, ep->mtu_size - pkt_entry->pkt_size, hmem_iface, device,
```
Comment (lrbison) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:2299]: I think the formatting is off here

File: prov/efa/src/rxr/rxr_op_entry.h [hunk 2]
```diff
@@ -80,6 +80,8 @@ struct rxr_atomic_ex {
 	int resp_iov_count;
 	struct iovec comp_iov[RXR_IOV_LIMIT];
 	int comp_iov_count;
+	void *result_desc;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_op_entry.h:83]: should be an array of `void *` to support multiple iov

File: prov/efa/src/rxr/rxr_atomic.c [hunk 3]
```diff
@@ -383,6 +380,8 @@ rxr_atomic_readwritemsg(struct fid_ep *ep,
 
 	ofi_ioc_to_iov(resultv, atomic_ex.resp_iov, result_count, datatype_size);
 	atomic_ex.resp_iov_count = result_count;
+	atomic_ex.result_desc = *result_desc;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_atomic.c:383]: `result_desc` is an array, and can have multiple entry. number of entry is in `recult_ioc`.

File: prov/efa/src/rxr/rxr_op_entry.h [hunk 4]
```diff
@@ -166,7 +168,7 @@ struct rxr_op_entry {
 	struct dlist_entry pending_recv_entry;
 #endif
 
-	size_t efa_outstanding_tx_ops;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_op_entry.h:169]: This format change is nice, but should not be in this commit

File: prov/efa/src/rxr/rxr_pkt_type_misc.c [hunk 5]
```diff
@@ -209,10 +209,10 @@ ssize_t rxr_pkt_init_cts(struct rxr_ep *ep,
 		cts_hdr->flags |= RXR_CTS_READ_REQ;
 
 	/* CTS is sent by different communication protocols.
-	 * CTS is sent using tx_entry in the emulated longcts read 
-	 * protocol. The emulated longcts write and the longcts 
+	 * CTS is sent using tx_entry in the emulated longcts read
+	 * protocol. The emulated longcts write and the longcts
 	 * message protocols sends CTS using rx_entry.
-	 * This check ensures appropriate tx_id and rx_id are
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:215]: Format change should be in a different commit

File: prov/efa/src/rxr/rxr_pkt_type_misc.c [hunk 6]
```diff
@@ -637,15 +637,23 @@ void rxr_pkt_handle_atomrsp_recv(struct rxr_ep *ep,
...
 	atomrsp_pkt = (struct rxr_atomrsp_pkt *)pkt_entry->pkt;
 	atomrsp_hdr = &atomrsp_pkt->hdr;
 	tx_entry = ofi_bufpool_get_ibuf(ep->op_entry_pool, atomrsp_hdr->recv_id);
 
-	ofi_copy_to_iov(tx_entry->atomic_ex.resp_iov,
-			tx_entry->atomic_ex.resp_iov_count,
-			0, atomrsp_pkt->data,
-			atomrsp_hdr->seg_length);
+    hmem_iface = ((struct efa_mr*) tx_entry->atomic_ex.result_desc)->peer.iface;
+    device = ((struct efa_mr*) tx_entry->atomic_ex.result_desc)->peer.device.reserved;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c:648]: identation wrong

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 7]
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
...
+{
+	char *host_data;
+	uint64_t device = efa_mr->peer.device.reserved;
+	int ret;
+
+	/* Step 1: Copy data from device to temporary host buffer */
+	host_data = malloc(iov->iov_len);
+	if (!host_data) {
+		return FI_ENOSPC;
+	}
```
Comment (lrbison) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: future optimization: use some sort of memory pool so we don't continually alloc/free host memory for this.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 8]
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
...
 
+static int rxr_cuda_fetch_atomic(struct efa_mr *efa_mr, struct iovec *iov, char* data, void* result, size_t dtsize, int op, int dt)
+{
+	char *host_data;
+	uint64_t device = efa_mr->peer.device.reserved;
+	int ret;
+
+	/* Step 1: Copy data from device to temporary host buffer */
+	host_data = malloc(iov->iov_len);
+	if (!host_data) {
```
Comment (darrylabbate) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: Probably worth it to wrap all the error checks in `OFI_UNLIKELY()`

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 9]
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
...
+		return FI_ENOSPC;
+	}
+
+	ret = ofi_copy_from_hmem(FI_HMEM_CUDA, device, host_data, iov->iov_base, iov->iov_len);
+	cuda_flush_rdma_operations();
+	if (!ret) {
+		return ret;
+	}
```
Comment (darrylabbate) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: Shouldn't `host_data` be `free`'d before returning?

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 10]
```diff
@@ -2417,13 +2489,50 @@ int rxr_pkt_proc_dc_write_rta(struct rxr_ep *ep,
...
+	                                      data,
+	                                      result,
+	                                      iov->iov_len / dtsize);
+
+	/* Step 3: Copy data from host buffer to device */
+	ret = ofi_copy_to_hmem(FI_HMEM_CUDA, device, iov->iov_base, host_data, iov->iov_len);
+	cuda_flush_rdma_operations();
+	if (!ret) {
+		return ret;
+	}
```
Comment (darrylabbate) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: Same here (`free` `host_data`?). Maybe turn these into a `goto` where `host_data` is `free`'d

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 7 above)
Comment (darrylabbate) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: I'm guessing `iov->iov_len` is dynamic (and we want to avoid VLAs). Otherwise it makes more sense to just have the string buffer be stack-allocated. E.g.
```c
char host_data[iov->iov_len];
```

Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: good idea.. switched to VLA's.

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 8 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: good idea.

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 9 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: no longer malloc'ed

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 10 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: no longer malloc'ed

File: prov/efa/src/rxr/rxr_op_entry.h (hunk 4 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_op_entry.h:169]: done

File: prov/efa/src/rxr/rxr_pkt_type_misc.c (hunk 5 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:215]: done

File: prov/efa/src/rxr/rxr_pkt_type_misc.c (hunk 6 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c:648]: done

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 11]
```diff
@@ -2227,8 +2229,14 @@ ssize_t rxr_pkt_init_rta(struct rxr_ep *ep, struct rxr_op_entry *tx_entry,
 
 	hdr_size = rxr_pkt_req_hdr_size_from_pkt_entry(pkt_entry);
 	data = (char *)pkt_entry->pkt + hdr_size;
-	data_size = ofi_copy_from_iov(data, ep->mtu_size - hdr_size,
-				      tx_entry->iov, tx_entry->iov_count, 0);
+	hmem_iface = ((struct efa_mr*) tx_entry->desc[0])->peer.iface;
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_pkt_type_req.c:2232]: `tx_entry->desc[0]` can be NULL for system memory

Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c:2232]: It will never be NULL.  Atomics are for one sided communication, and in order to use 1 sided communication, you must register all your memory.

Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c:2232]: The compare, result, and buff must all be registered no-matter the memory type in order to use libfabric atomics AFAIK... I think an assert would be better here if you are concerned.

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 1 above)
Comment (a-szegel) [reply prov/efa/src/rxr/rxr_pkt_type_req.c:2299]: ty fixed.
